README.md
# Geotrigger Editor

The Geotrigger Editor is a client-side web application for creating and editing Geotrigger rules.

This is the open source version of the Geotrigger Editor available for Esri customers on the ArcGIS for Developers site. Read more about the editor [here](https://developers.arcgis.com/geotrigger-service/guide/geotrigger-editor/), and find out more about the Esri Geotrigger Service [here](https://developers.arcgis.com/en/features/geotrigger-service/).

## Features

* **Draw** trigger boundaries on a map
* **List** all of an application's triggers
* **Create**, **Edit** and **Delete** trigger conditions, actions, and boundaries
* **Search** for triggers by tag or ID

## Instructions

### Installation

Tagged releases include a `dist` folder which contains all the files you need to include the Geotrigger Editor in your project.

* The built JavaScript and CSS files are available in both expanded and minified versions in `dist/js` and `dist/css` respectively.
* All images used by the Editor are in `dist/img`.
* The files in `examples/default` show the easiest way to get the editor running in a browser.
* If you need to support IE 8/9, you're going to need a server-side [proxy](#proxy) to communicate with ArcGIS Online and the Geotrigger server. The files in `examples/proxy` show a quick way to do this using node.js.

**Dependencies**:

The Geotrigger Editor has some external dependencies:

* jQuery [1.10.2](https://github.com/jquery/jquery/releases/tag/1.10.2) (CDN: [Google](//ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js), [jQuery](http://codeorigin.jquery.com/jquery-1.10.2.min.js))
* Underscore [1.5.2](https://github.com/jashkenas/underscore/releases/tag/1.5.2) (CDN: [cdnjs](http://cdnjs.com/libraries/underscore.js/))
* Backbone [1.1.0](https://github.com/jashkenas/backbone/releases/tag/1.1.0) (CDN: [cdnjs](http://cdnjs.com/libraries/backbone.js/))
* Backbone.Marionette [1.4.1-bundled](https://github.com/marionettejs/backbone.marionette/releases/tag/v1.4.1) (CDN: [cdnjs](http://cdnjs.com/libraries/backbone.marionette/))
* Handlebars [1.3.0](https://github.com/wycats/handlebars.js/releases/tag/v1.3.0) (runtime only) (CDN: [cdnjs](http://cdnjs.com/libraries/handlebars.js/))
* Leaflet [0.7.7](https://github.com/Leaflet/Leaflet/releases/tag/v0.7.7) (CDN: [leaflet](http://leafletjs.com/download.html), [cdnjs](http://cdnjs.com/libraries/leaflet/))
* Leaflet.draw [0.2.3](https://github.com/Leaflet/Leaflet.draw/releases/tag/0.2.3)
* Geotrigger.js [1.0.0](https://github.com/Esri/geotrigger-js)
* Esri-Leaflet [1.0.1](https://github.com/Esri/esri-leaflet/releases/tag/v1.0.1) (CDN: [jsdelivr](http://www.jsdelivr.com/projects/leaflet.esri))
* Esri-Leaflet Geocoder [1.0.2](https://github.com/Esri/esri-leaflet-geocoder/releases/tag/v1.0.2) (CDN: [jsdelivr](http://www.jsdelivr.com/projects/leaflet.esri.geocoder))

We recommend loading dependencies from a CDN when possible. All dependencies are hosted on CDNs with the exception of Geotrigger.js, which you'll need to host yourself.

If you plan to support legacy browsers (IE 8), you'll also need [html5shiv](https://github.com/aFarkas/html5shiv) and [json2](https://github.com/douglascrockford/JSON-js), which can be included with a conditional comment like so:

```html
<!--[if lte IE 8]>
  <script src="//cdnjs.cloudflare.com/ajax/libs/json2/20121008/json2.min.js"></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/r29/html5.min.js"></script>
<![endif]-->
```

You can take a look at `examples/default` to see how to easily include all the necessary files.

**Browser Support**:

The Geotrigger Editor is tested against the latest versions of Chrome, Firefox, Safari, and Internet Explorer versions 8, 9, and 10. IE 8 & 9 only work with a [proxy](#proxy).

---

### Starting the Editor

The Editor is a [Backbone.Marionette](https://github.com/marionettejs/backbone.marionette) application namespaced as `Geotrigger.Editor`. The only method you need to start the application is [`.start`](https://github.com/marionettejs/backbone.marionette/blob/master/docs/marionette.application.md#starting-an-application).

### `Geotrigger.Editor.start(options)`

```js
Geotrigger.Editor.start({
  session: { // required
    clientId: 'XXXXXX', // required
    clientSecret: 'XXXXXX', // required
    proxy: '/path/to/proxy' // required if you need to support IE 8/9
  }
});
```

The Editor requires an empty `#gt-editor` element in the DOM to attach to, e.g. `<div id="gt-editor"></div>`. See `examples/default` for an example.

**Configuration Options**

The `options` parameter is an object that can contain the following configuration options:

#### `session`

The `session` object is required to authenticate with ArcGIS and interact with the Geotrigger Service. It has two required properties:

* `clientId`: your ArcGIS application's client ID
* `clientSecret`: your ArcGIS application's client secret

The client ID and secret are required in order for the Geotrigger Service to grant editing rights to the user for the associated application.

It can take all of the same properties as those outlined in the geotrigger.js [config options](https://github.com/Esri/geotrigger-js#config-options).

#### more

There are a lot more configuration options, look at `src/js/modules/config.js` to see them all.

---

### Proxy

The Geotrigger Editor is a client-side only application that relies on [Cross Origin Resource Sharing](https://developer.mozilla.org/en-US/docs/HTTP/Access_control_CORS) to communicate with the Geotrigger Service.

**If you need to serve browsers that don't support CORS, you're going to need a proxy.** This means Internet Explorer 8 and 9 in particular -- see [caniuse.com/cors](http://caniuse.com/cors). Thankfully Internet Explorer 10 supports CORS along with all the latest versions of modern browsers (Chrome, Firefox, Safari).

#### Client

On the client side, just be sure to start the editor with a path to the proxy, like so:

```js
Geotrigger.Editor.start({
  session: { // required
    clientId: 'XXXXXX', // required
    clientSecret: 'XXXXXX', // required
    proxy: '/proxy/'
  }
});
```

The `proxy` option should be an absolute path to the proxy server endpoint (starts with `/`). For the proxy to work it must be on the same domain.

#### Server

On the server side, you'll need a working proxy to forward API requests to `geotrigger.arcgis.com/*` and `arcgis.com/sharing/oauth2/*` and return the response back to the browser. `/examples/proxy/` shows how to do this using Node.js.

---

### Developing & Building

To run the development environment, you will need the following:

1. [Node.js](http://nodejs.org/)
1. [Grunt](http://gruntjs.com/)
1. [Ruby](https://www.ruby-lang.org/en/downloads/) and [Compass](http://compass-style.org/install/) (to compile SASS)

Once [Node.js](https://nodejs.org/en/download/) is installed, you can install the Grunt command line interface by running `npm install -g grunt-cli`. This will install the grunt-cli package locally ([reference](https://npmjs.org/doc/install.html)).

#### Local Setup

1. Clone the repository and `cd` into the `geotrigger-editor` directory.
2. Run `npm install` to lay down the  project's dependencies.
3. Run `grunt` to compile the source code and spin up a local server which watches the source code for changes.

#### Working locally

You'll need an HTTP server to serve files from the root of the repository. The `grunt dev` task takes care of building all source files to the `dist` folder (`dist/js/geotrigger-editor.js` and `dist/css/geotrigger-editor.css`), watching local files for changes, and running a local server for you at http://localhost:8080 (see the `Grunt Tasks` section for more information).

#### Testing locally (todo)

Testing requires PhantomJS to be running. You can install it with homebrew (`brew install phantomjs`), and get it started by running `phantomjs --webdriver=4444`. The local server should already be running too. The `grunt test` task takes care of the latter part (see `Grunt Tasks` below).

#### Grunt Tasks

##### 1. `grunt dev`

This task will rebuild temporary files for development, run a server at `localhost:8080`, then continuously watch for changes in the `src` directory until you end the process. The `index.html` file in the root of the repository is already set up to use these files.

##### 2. `grunt test` **not yet implemented**

This task will start a test server at `localhost:8081`, then run the jshint, complexity, and cucumber tasks to see if the code does not smell.

##### 3. `grunt build`

This task will build the production version of the editor into the `dist` folder.

##### 4. `grunt build_img`

This task will build the production version of the editor's image assets into the `dist` folder. This task has been broken out of the main `build` task because [smushit](https://github.com/heldr/grunt-smushit) takes a long time.

## Issues

Find a bug or want to request a new feature? Please let us know by submitting an [issue](https://github.com/Esri/geotrigger-editor/issues).

## Contributing

Esri welcomes contributions from anyone and everyone. Please see our [guidelines for contributing](https://github.com/esri/contributing).

## Licensing

Copyright 2015 Esri

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

> http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

A copy of the license is available in the repository's [license.txt](https://raw.github.com/Esri/geotrigger-editor/master/LICENSE) file.

[](Esri Tags: Geotrigger Editor Geolocation Web Mobile Browser HTML5)
[](Esri Language: JavaScript)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]

## [0.2.2] - 2016-1-21

* bump `leaflet` to `0.7.7` and move to CDN
* bump `esri-leaflet` to [`1.0.1`](https://github.com/Esri/esri-leaflet/releases/tag/v1.0.1) and move to CDN
* bump `esri-leaflet-geocoder` to [`1.0.2`](https://github.com/Esri/esri-leaflet-geocoder/releases/tag/v1.0.2) and move to CDN
* bump `geotrigger-js` to [`1.0.4`](https://github.com/Esri/geotrigger-js/releases/tag/v1.0.4) and move to CDN
* move `Leaflet.draw` to CDN
* changed basemap to [Esri World Topographic](http://www.arcgis.com/home/item.html?id=30e5fe3149c34df1ba922e6f5bbf808f)
* moved geosearch control below the basemap widget
* corrected relative path links to images in geotrigger-editor.css
* bumped `grunt-contrib-uglify` version to `~0.10.0` to resolve a bug compiling the built source
* corrected relative path links to images
* stopped tracking the `dist` directory and got rid of the `tmp` directory entirely
* admitted that our current proxy sample still needs work

## [0.2.1] - 2014-4-10
### Fixed
* Fix IE scrollbar CSS bug
* Add warning notifying users of lack of support for touch-enabled browsers
* Fix underscoreDeepExtend breaking in AMD environments
* bump leaflet.draw to 0.2.3
* bump geotrigger-js to 1.0.0
* bump esri-leaflet-geocoder to 0.0.1-beta.3
* Add basemap switcher ([#160](https://github.com/Esri/geotrigger-editor/issues/160))

## [0.2.0] - 2014-2-11
### Fixed
* bump leaflet to 0.7.1
* bump handlebars to 1.3.0
* bump marionette to 1.4.1
* remove marionette from `vendor` and `.gitmodules`
* fix CSS rules leaking into global scope
* fix bug where callback URL & tracking profile fields were incorrectly filled after trigger/update
* add geocoder control for fast map traversal ([#195](https://github.com/Esri/geotrigger-editor/issues/195))
* use compass `image-url` to allow easily generating URLs for different structures using grunt-compass and `httpImagesPath`
* change image folder from `img/icons` to `img/geotrigger-editor` to avoid conflicts
* move namespace to `Geotrigger.Editor` ([#185](https://github.com/Esri/geotrigger-editor/issues/185))
* add support for `properties`, `rateLimit`, and `times` trigger parameters ([#192](https://github.com/Esri/geotrigger-editor/issues/192) & [#193](https://github.com/Esri/geotrigger-editor/issues/193))
* fix for duplicate triggers appearing in list on trigger/create overwrite ([#189](https://github.com/Esri/geotrigger-editor/issues/189) & [#194](https://github.com/Esri/geotrigger-editor/issues/194))

## [0.1.3] - 2014-1-6
### Fixed
* fix CSS issue causing IDs not to display in title area on edit ([#186](https://github.com/Esri/geotrigger-editor/issues/186))
* bump esri-leaflet to v0.0.1-beta.3
* bump geotrigger-js to v0.1.5
* fix highlight issue for list items
* make some routes (list, new, edit) more forgiving with trailing slashes
* add beautifier tasks (`grunt verify` & `grunt format`), format js & templates

## [0.1.2] - 2013-12-31
### Fixed
* better template compression
* trigger form validation
* proper JSON parsing for notification data ([#180](https://github.com/Esri/geotrigger-editor/issues/180))
* replace optional title with ability to enter custom trigger ID ([#179](https://github.com/Esri/geotrigger-editor/issues/179))

## [0.1.1] - 2013-12-11
### Fixed
* wrap trigger IDs in search link on list ([#168](https://github.com/Esri/geotrigger-editor/issues/168))
* check for distance before geojson when rendering shapes on map ([#175](https://github.com/Esri/geotrigger-editor/pull/175))
* remove compiled.js from source control ([#176](https://github.com/Esri/geotrigger-editor/pull/176))
* update tag label and placeholder text in form ([#176](https://github.com/Esri/geotrigger-editor/pull/176))
* wrap application in a closure ([#176](https://github.com/Esri/geotrigger-editor/pull/176))

## 0.1.0 - 2013-11-19

This is the first public release of the Geotrigger Editor for the ArcGIS Geotrigger Service.

Currently the Geotrigger Editor is in development and should be thought of as a beta or preview.

[Unreleased]: https://github.com/Esri/geotrigger-editor/compare/v0.2.1...HEAD
[0.2.2]: https://github.com/Esri/geotrigger-editor/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/Esri/geotrigger-editor/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/Esri/geotrigger-editor/compare/v0.1.3...v0.2.0
[0.1.3]: https://github.com/Esri/geotrigger-editor/compare/v0.1.2...v0.1.3
[0.1.2]: https://github.com/Esri/geotrigger-editor/compare/v0.1.1...v0.1.2
[0.1.1]: https://github.com/Esri/geotrigger-editor/compare/v0.1.0...v0.1.1


Gruntfile.js
module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),

    connect: {
      dist: {
        options: {
          port: 8080,
          base: '.',
          hostname: '*'
        }
      },
      test: {
        options: {
          port: 8081,
          base: '.',
          hostname: '*'
        }
      }
    },

    uglify: {
      options: {
        banner: '/*! <%= pkg.name %> - v<%= pkg.version %> - <%= grunt.template.today("yyyy-mm-dd") %>\n' +
        '*   <%= pkg.homepage %>\n' +
        '*   Copyright (c) <%= grunt.template.today("yyyy") %> Environmental Systems Research Institute, Inc.\n' +
        '*   Apache 2.0 License */\n\n',
        report: 'gzip',
        enclose: {}
      },
      dist: {
        files: {
          'dist/js/<%= pkg.name %>.min.js': [
            'src/js/app.js',
            'tmp/templates.js',
            'src/js/lib/*.js',
            'src/js/modules/*.js',
            'src/js/controllers/*.js',
            'src/js/models/*.js',
            'src/js/views/*.js'
          ]
        }
      }
    },

    watch: {
      scripts: {
        files: [
          'src/js/controllers/*.js',
          'src/js/models/*.js',
          'src/js/modules/*.js',
          'src/js/views/*.js',
          'src/js/app.js',
          'src/scss/**/*.scss',
          'src/img/**/*.jpg',
          'src/img/**/*.png',
          'src/templates/**/*.hbs'
        ],
        tasks: [
          'jshint',
          'clean:dist',
          'compass:dist',
          'handlebars',
          'concat:dist'
        ]
        // options: {
        //   nospawn: true
        // }
      }
    },

    compass: {
      // Production
      distmin: {
        options: {
          sassDir: 'src/scss',
          cssDir: 'dist/css',
          imagesDir: 'dist/img/geotrigger-editor',
          httpImagesPath: '../img/geotrigger-editor',
          environment: 'production',
          raw: 'asset_cache_buster :none\n'
        }
      },
      dist: {
        options: {
          sassDir: 'src/scss',
          cssDir: 'dist/css',
          imagesDir: 'dist/img/geotrigger-editor',
          httpImagesPath: '../img/geotrigger-editor',
          noLineComments: true,
          raw: 'asset_cache_buster :none\n'
        }
      }
    },

    rename: {
      distmin: {
        src: 'dist/css/geotrigger-editor.css',
        dest: 'dist/css/geotrigger-editor.min.css'
      }
    },

    jshint: {
      options: {
        'browser': true,
        'curly': true,
        'eqnull': true,
        'evil': true,
        'expr': true,
        'multistr': true,
        'trailing': true,
        'undef': false,
        'wsh': true,
        'sub': true,
        'boss': true
      },
      files: [
        'src/js/controllers/*.js',
        'src/js/models/*.js',
        'src/js/modules/*.js',
        'src/js/views/*.js',
        'src/js/app.js'
      ]
    },

    concat: {
      options: {
        banner: '/*! <%= pkg.name %> - v<%= pkg.version %> - <%= grunt.template.today("yyyy-mm-dd") %>\n' +
        '*   <%= pkg.homepage %>\n' +
        '*   Copyright (c) <%= grunt.template.today("yyyy") %> Environmental Systems Research Institute, Inc.\n' +
        '*   Apache 2.0 License */\n\n',
        // define a string to put between each file in the concatenated output
        separator: '\n\n'
      },
      dist: {
        files: {
          'dist/js/<%= pkg.name %>.js': [
            'src/js/header.js',
            'src/js/app.js',
            'tmp/templates.js',
            'src/js/lib/*.js',
            'src/js/modules/*.js',
            'src/js/controllers/*.js',
            'src/js/models/*.js',
            'src/js/views/*.js',
            'src/js/footer.js'
          ]
        }
      }
    },

    complexity: {
      generic: {
        src: [
          'src/js/controllers/*.js',
          'src/js/models/*.js',
          'src/js/modules/*.js',
          'src/js/views/*.js',
          'src/js/app.js',
          'tasks/grunt-complexity.js'
        ],
        options: {
          // jsLintXML: 'report.xml', // create XML JSLint-like report
          // checkstyleXML: 'checkstyle.xml', // create checkstyle report
          errorsOnly: false, // show only maintainability errors
          cyclomatic: 15,
          halstead: 30,
          maintainability: 80
        }
      }
    },

    cucumberjs: {
      files: 'features',
      options: {
        steps: 'features/step_definitions',
        format: 'pretty'
      }
    },

    clean: {
      dist: {
        src: [
          'dist/css/',
          'dist/js/'
        ]
      },
      img: {
        src: [
          'dist/img/'
        ]
      }
    },

    handlebars: {
      compile: {
        options: {
          namespace: 'Geotrigger.Editor.Templates',
          processContent: function(content, filepath) {
            content = content.replace(/^[\x20\t]+/mg, '').replace(/[\x20\t]+$/mg, '');
            content = content.replace(/^[\r\n]+/, '').replace(/[\r\n]*$/, '\n');
            return content;
          },
          processName: function(filePath) {
            console.log(filePath);
            var process = filePath.split('src/templates/')[1];
            return process.split('.hbs')[0];
          }
        },
        files: {
          'tmp/templates.js': 'src/templates/**/*.hbs'
        }
      }
    },

    smushit: {
      // filter by filetype
      dist: {
        src: [
          'src/img/**/*.png',
          'src/img/**/*.jpg'
        ],
        dest: 'dist/img/geotrigger-editor'
      }
    },

    jsbeautifier: {
      verify: {
        files: {
          src: [
            'src/js/**/*.js', '!src/js/lib/*.*', '!src/js/geotrigger-editor.js',
            'src/templates/*.hbs'
          ]
        },
        options: {
          mode: 'VERIFY_ONLY',
          html: {
            fileTypes: ['.hbs'],
            braceStyle: 'collapse',
            indentChar: ' ',
            indentScripts: 'keep',
            indentSize: 2,
            maxPreserveNewlines: 2,
            preserveNewlines: true,
            unformatted: ['a', 'sub', 'sup', 'b', 'i', 'u', 'span', 'strong', 'em'],
            wrapLineLength: 0
          },
          js: {
            braceStyle: 'collapse',
            breakChainedMethods: false,
            e4x: false,
            evalCode: false,
            indentChar: " ",
            indentLevel: 0,
            indentSize: 2,
            indentWithTabs: false,
            jslintHappy: true,
            keepArrayIndentation: false,
            keepFunctionIndentation: false,
            maxPreserveNewlines: 2,
            preserveNewlines: true,
            spaceBeforeConditional: true,
            spaceInParen: false,
            unescapeStrings: false,
            wrapLineLength: 0
          }
        }
      },
      format: {
        files: {
          src: [
            'src/js/**/*.js', '!src/js/lib/*.*', '!src/js/geotrigger-editor.js',
            'src/templates/*.hbs'
          ]
        },
        options: {
          mode: 'VERIFY_AND_WRITE',
          html: {
            fileTypes: ['.hbs'],
            braceStyle: 'collapse',
            indentChar: ' ',
            indentScripts: 'keep',
            indentSize: 2,
            maxPreserveNewlines: 2,
            preserveNewlines: true,
            unformatted: ['a', 'sub', 'sup', 'b', 'i', 'u', 'span', 'strong', 'em'],
            wrapLineLength: 0
          },
          js: {
            braceStyle: 'collapse',
            breakChainedMethods: false,
            e4x: false,
            evalCode: false,
            indentChar: " ",
            indentLevel: 0,
            indentSize: 2,
            indentWithTabs: false,
            jslintHappy: true,
            keepArrayIndentation: false,
            keepFunctionIndentation: false,
            maxPreserveNewlines: 2,
            preserveNewlines: true,
            spaceBeforeConditional: true,
            spaceInParen: false,
            unescapeStrings: false,
            wrapLineLength: 0
          }
        }
      }
    }
  });

  // load npm tasks
  require('matchdep').filterDev('grunt-*').forEach(grunt.loadNpmTasks);

  // Tasks
  grunt.registerTask('test', [
    'connect:test',
    'jshint',
    'complexity',
    'cucumberjs'
  ]);

  grunt.registerTask('verify', [
    'jshint',
    'complexity',
    'jsbeautifier:verify'
  ]);

  grunt.registerTask('format', [
    'jsbeautifier:format'
  ]);

  grunt.registerTask('dev', [
    'jshint',
    'clean:dist',
    'compass:dist',
    'handlebars',
    'concat:dist',
    'build_img',
    'connect:dist',
    'watch'
  ]);

  grunt.registerTask('build', [
    'clean:dist',
    'compass:distmin',
    'rename:distmin',
    'compass:dist',
    'handlebars',
    'concat:dist',
    'build_img',
    'uglify:dist'
  ]);

  grunt.registerTask('build_img', [
    'clean:img',
    'smushit'
  ]);

  grunt.registerTask('default', [
    'dev'
  ]);
};
